### PR TITLE
Add support for networking, encryption and replication params in GCP secret-sync destination

### PR DESCRIPTION
**Pull request by siyer-corp**
### Description

Add support for 7 new fields to the `vault_secrets_sync_gcp_destination` resource for enhanced networking, encryption, and multi-region replication capabilities.
  - `allowed_ipv4_addresses` - IP-based access control for synced secrets (Requires Vault 1.19+)
  - `allowed_ipv6_addresses` - IPv6 CIDR block support (Requires Vault 1.19+)
  - `allowed_ports` - Port-based access restrictions (Requires Vault 1.19+)
  - `disable_strict_networking` - Optional strict networking bypass (Requires Vault 1.19+)
  - `global_kms_key` - Single KMS key for all regions (Requires Vault 1.19+)
  - `locational_kms_keys` - Per-region KMS key configuration (Requires Vault 1.19+)
  - `replication_locations` - Multi-region secret replication support (Requires Vault 1.18+)

### Checklist
- [x] Added [CHANGELOG](https://github.com/hashicorp/terraform-provider-vault/blob/master/CHANGELOG.md) entry (only for user-facing changes)
- [x] Acceptance tests where run against all supported Vault Versions

### Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccXXX'

...
```

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

## PCI review checklist

- [ ] I have documented a clear reason for, and description of, the change I am making.

- [ ] If applicable, I've documented a plan to revert these changes if they require more than reverting the pull request.

- [ ] If applicable, I've documented the impact of any changes to security controls.

  Examples of changes to security controls include using new access control methods, adding or removing logging pipelines, etc.

### CODE REVIEW

**Comment by vijayavelsekar** [comment]
@siyer-corp Acceptance testing screenshot missing.

File: vault/resource_secrets_sync_gcp_destination.go [hunk 1]
```diff
@@ -18,19 +18,51 @@ const (
...
+
+	if provider.IsAPISupported(meta, provider.VaultVersion119) {
+		fields = append(fields, []string{
+			consts.FieldAllowedIPv4Addresses,
+			consts.FieldAllowedIPv6Addresses,
+			consts.FieldAllowedPorts,
+			consts.FieldDisableStrictNetworking,
+			consts.FieldLocationalKmsKeys,
+			consts.FieldGlobalKmsKey,
+			consts.FieldReplicationLocations,
```
Comment (vijayavelsekar) [inline vault/resource_secrets_sync_gcp_destination.go]: [replication_locations](https://developer.hashicorp.com/vault/api-docs/v1.18.x/system/secrets-sync#replication_locations) is supported from v1.18.0. So need to update the validations accordingly, here and in the acceptance testing as well.

File: vault/resource_secrets_sync_gcp_destination.go [hunk 2]
```diff
@@ -62,16 +94,59 @@ func gcpSecretsSyncDestinationResource() *schema.Resource {
 				ForceNew:    true,
 				Description: "The target project to manage secrets in.",
 			},
+			consts.FieldAllowedIPv4Addresses: {
+				Type:        schema.TypeList,
+				Optional:    true,
+				Description: "Allowed IPv4 addresses for network connectivity.",
```
Comment (vijayavelsekar) [inline vault/resource_secrets_sync_gcp_destination.go]: You can also mention that it should be in CIDR-notation

File: vault/resource_secrets_sync_gcp_destination.go [hunk 3]
```diff
@@ -62,16 +94,59 @@ func gcpSecretsSyncDestinationResource() *schema.Resource {
...
+				Elem:        &schema.Schema{Type: schema.TypeString},
+			},
+			consts.FieldAllowedIPv6Addresses: {
```
Comment (vijayavelsekar) [inline vault/resource_secrets_sync_gcp_destination.go:110]: You can also mention that it should be in CIDR-notation

File: vault/resource_secrets_sync_gcp_destination.go (hunk 1 above)
Comment (siyer-corp) [reply vault/resource_secrets_sync_gcp_destination.go]: Added the version check for replication_locations and added a test specific to it. Acceptance test screenshots are attached to the JIRA instead.

File: vault/resource_secrets_sync_gcp_destination.go (hunk 2 above)
Comment (siyer-corp) [reply vault/resource_secrets_sync_gcp_destination.go]: Added more context in the description.

File: vault/resource_secrets_sync_gcp_destination.go [hunk 4]
```diff
@@ -62,16 +100,59 @@ func gcpSecretsSyncDestinationResource() *schema.Resource {
 				ForceNew:    true,
 				Description: "The target project to manage secrets in.",
 			},
+			consts.FieldAllowedIPv4Addresses: {
+				Type:        schema.TypeList,
+				Optional:    true,
+				Description: "Allowed IPv4 addresses for outbound network connectivity in CIDR notation. If not set, all IPv6 addresses are allowed.",
```
Comment (vijayavelsekar) [inline vault/resource_secrets_sync_gcp_destination.go]: 1. Typo: replace IPv6 to IPv4
2. nit: Longer description lines can be broken into 2 lines

File: vault/resource_secrets_sync_gcp_destination_test.go [hunk 5]
```diff
@@ -92,3 +93,181 @@ resource "vault_secrets_sync_gcp_destination" "test" {
 
 	return ret
 }
+
+func TestGCPSecretsSyncDestination_NetworkingAndEncryption(t *testing.T) {
```
Comment (vijayavelsekar) [inline vault/resource_secrets_sync_gcp_destination_test.go]: Based on the previous comment that I got from Ben, I was having similar thought here. Let me know if this works.
Instead of individual test cases for Networking, Encryption, Replication and ReplicationLocation. Can we have them all in single test function with multiple steps that test all these scenarios ?
Also I think it would be good to have  a separate test function for negative test cases alone (example: Invalid CIDR notation, Conflicting encryption configurations, ....etc).
Let me know your thoughts.

Comment (siyer-corp) [reply vault/resource_secrets_sync_gcp_destination_test.go]: That makes sense, have addressed these.

File: vault/resource_secrets_sync_gcp_destination.go [hunk 6]
```diff
@@ -62,16 +100,68 @@ func gcpSecretsSyncDestinationResource() *schema.Resource {
...
+			},
+			consts.FieldAllowedIPv6Addresses: {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Description: "Allowed IPv6 addresses for outbound network connectivity in CIDR notation. " +
+					"If not set, all IPv6 addresses are allowed.",
+				Elem: &schema.Schema{Type: schema.TypeString},
+			},
+			consts.FieldAllowedPorts: {
+				Type:     schema.TypeSet,
```
Comment (fairclothjm) [inline vault/resource_secrets_sync_gcp_destination.go:118]: Are these TypeSet fields being converted to a list when calling the Vault API?

Comment (siyer-corp) [reply vault/resource_secrets_sync_gcp_destination.go:118]: Yes these will be passed to SyncDestinationCreateUpdateWithOptions function in the secrets_sync.go where it'll manage the conversion. Added a comment on line 153 for this context

**Review by fairclothjm — DISMISSED** [review]
LGTM
